File: pkg/client/durable_ordered_release.go
package client

import (
	"fmt"
	"time"

	v1 "github.com/hatchet-dev/hatchet/internal/services/shared/proto/v1"
)

const (
	defaultParkTimeout = 5 * time.Second
	defaultGapTimeout  = 60 * time.Second
	gateSweepInterval  = 500 * time.Millisecond
)

// OrderedReplayGapError is returned to waiters when a gap in the satisfied
// order sequence persists beyond the gap timeout. This indicates the recorded
// history diverged from the current code (e.g. a deploy removed a durable
// await mid-flight), since a recorded completion that should fill the gap was
// never re-awaited.
type OrderedReplayGapError struct {
	TaskExternalID  string
	HeldOrders      []int64
	MissingOrder    int64
	InvocationCount int32
}

func (e *OrderedReplayGapError) Error() string {
	return fmt.Sprintf(
		"non-determinism detected for task %s (invocation %d): completion with satisfied order %d was never delivered while later completions %v arrived; the recorded history likely diverged from the current code",
		e.TaskExternalID, e.InvocationCount, e.MissingOrder, e.HeldOrders,
	)
}

// invocationGate serializes the release of ordered EntryCompleted responses for
// a single durable task invocation. Completions are released to user code in
// satisfied_order, and after a release wakes a parked continuation, further
// releases are held until that continuation parks again (registers its next
// awaited entry), the runtime signals quiescence, or the park timeout elapses.
type invocationGate struct {
	wakeSince time.Time
	gapSince  time.Time
	held      map[int64]*v1.DurableTaskResponse
	released  int64
	wakes     int
}

func (l *DurableTaskListener) gateFor(key PendingAckKey) *invocationGate {
	g, ok := l.gates[key]
	if !ok {
		g = &invocationGate{held: make(map[int64]*v1.DurableTaskResponse)}
		l.gates[key] = g
	}
	return g
}

// handleOrderedEntryCompleted routes an EntryCompleted response carrying a
// satisfied order through the invocation's gate. Must NOT be called while
// holding gateMu.
func (l *DurableTaskListener) handleOrderedEntryCompleted(key PendingAckKey, order int64, resp *v1.DurableTaskResponse) {
	l.gateMu.Lock()
	g := l.gateFor(key)

	if order <= g.released {
		// re-delivery of an already-released completion (e.g. after reconnect):
		// bypass the gate.
		l.gateMu.Unlock()
		l.deliverCompletion(resp)
		return
	}

	g.held[order] = resp
	l.pumpLocked(key, g)
	l.gateMu.Unlock()
}

// pumpLocked releases contiguously ordered completions while the gate is open.
// Callers must hold gateMu.
func (l *DurableTaskListener) pumpLocked(key PendingAckKey, g *invocationGate) {
	for g.wakes == 0 {
		resp, ok := g.held[g.released+1]
		if !ok {
			break
		}

		delete(g.held, g.released+1)
		g.released++

		if l.deliverCompletion(resp) {
			// the release woke a parked continuation: hold further releases
			// until it parks again.
			g.wakes++
			g.wakeSince = time.Now()
		}
		// if nobody was waiting, the completion was buffered for a continuation
		// that is still running; keep pumping so a parked continuation awaiting
		// a later order is not deadlocked.
	}

	if len(g.held) > 0 && g.wakes == 0 {
		if g.gapSince.IsZero() {
			g.gapSince = time.Now()
		}
	} else {
		g.gapSince = time.Time{}
	}
}

// deliverCompletion hands an EntryCompleted response to a registered waiter,
// or buffers it for late registration. Returns true if a parked continuation
// was woken.
func (l *DurableTaskListener) deliverCompletion(resp *v1.DurableTaskResponse) bool {
	completed := resp.GetEntryCompleted()
	ref := completed.GetRef()
	key := PendingCallbackKey{
		TaskID:    ref.GetDurableTaskExternalId(),
		SignalKey: int64(ref.GetInvocationCount()),
		BranchID:  ref.GetBranchId(),
		NodeID:    ref.GetNodeId(),
	}

	l.pendingCallbacksMu.Lock()
	ch, ok := l.pendingCallbacks[key]
	if ok {
		delete(l.pendingCallbacks, key)
	}
	l.pendingCallbacksMu.Unlock()

	if ok {
		select {
		case ch <- CallbackResult{Resp: resp}:
		default:
		}
		return true
	}

	l.bufferedCompletionsMu.Lock()
	l.bufferedCompletions[key] = resp
	l.bufferedCompletionsMu.Unlock()

	return false
}

// notifyParked records that a continuation of the given invocation parked
// (registered its next awaited entry without a buffered result), opening the
// gate for the next ordered release.
func (l *DurableTaskListener) notifyParked(key PendingAckKey) {
	l.gateMu.Lock()
	defer l.gateMu.Unlock()

	g, ok := l.gates[key]
	if !ok {
		return
	}

	if g.wakes > 0 {
		g.wakes--
	}

	l.pumpLocked(key, g)
}

// NotifyInvocationQuiesced signals that the durable task function for the given
// invocation returned (or otherwise has no running continuations), releasing
// any gate held on its behalf.
func (l *DurableTaskListener) NotifyInvocationQuiesced(taskExternalID string, invocationCount int32) {
	key := PendingAckKey{TaskID: taskExternalID, SignalKey: int64(invocationCount)}

	l.gateMu.Lock()
	defer l.gateMu.Unlock()

	g, ok := l.gates[key]
	if !ok {
		return
	}

	g.wakes = 0
	l.pumpLocked(key, g)
}

// sweepGates enforces the park and gap timeouts. Park timeout: a woken
// continuation that never parks (e.g. performed unrecorded blocking work or
// returned without the runtime quiesce hook) forces the gate open with a loud
// warning. Gap timeout: a persistent hole in the order sequence while later
// completions are held fails the invocation's waiters with an
// OrderedReplayGapError instead of hanging.
func (l *DurableTaskListener) sweepGates() {
	type gapFailure struct {
		err *OrderedReplayGapError
		key PendingAckKey
	}

	var failures []gapFailure

	l.gateMu.Lock()
	now := time.Now()

	for key, g := range l.gates {
		if g.wakes > 0 && now.Sub(g.wakeSince) > l.parkTimeout {
			l.l.Warn().
				Str("task_id", key.TaskID).
				Int64("invocation_count", key.SignalKey).
				Msgf("DurableTaskListener: continuation did not park within %s after a gated release; forcing the completion gate open. durable task code should not perform unrecorded blocking work between durable operations", l.parkTimeout)
			g.wakes = 0
			l.pumpLocked(key, g)
		}

		if len(g.held) > 0 && g.wakes == 0 && !g.gapSince.IsZero() && now.Sub(g.gapSince) > l.gapTimeout {
			heldOrders := make([]int64, 0, len(g.held))
			for o := range g.held {
				heldOrders = append(heldOrders, o)
			}

			failures = append(failures, gapFailure{
				key: key,
				err: &OrderedReplayGapError{
					TaskExternalID:  key.TaskID,
					InvocationCount: int32(key.SignalKey), // nolint:gosec
					MissingOrder:    g.released + 1,
					HeldOrders:      heldOrders,
				},
			})

			delete(l.gates, key)
		}
	}
	l.gateMu.Unlock()

	for _, f := range failures {
		l.l.Error().
			Str("task_id", f.key.TaskID).
			Int64("invocation_count", f.key.SignalKey).
			Msg(f.err.Error())

		l.failInvocationWaiters(f.key, f.err)
	}
}

// failInvocationWaiters delivers an error to every pending callback and event
// ack belonging to the given invocation.
func (l *DurableTaskListener) failInvocationWaiters(key PendingAckKey, err error) {
	l.pendingCallbacksMu.Lock()
	for k, ch := range l.pendingCallbacks {
		if k.TaskID == key.TaskID && k.SignalKey == key.SignalKey {
			delete(l.pendingCallbacks, k)
			select {
			case ch <- CallbackResult{Err: err}:
			default:
			}
		}
	}
	l.pendingCallbacksMu.Unlock()

	l.pendingEventAcksMu.Lock()
	if ch, ok := l.pendingEventAcks[key]; ok {
		delete(l.pendingEventAcks, key)
		select {
		case ch <- EventAckResult{Err: err}:
		default:
		}
	}
	l.pendingEventAcksMu.Unlock()
}

func (l *DurableTaskListener) runGateSweeper(stop <-chan struct{}) {
	ticker := time.NewTicker(gateSweepInterval)
	defer ticker.Stop()

	for {
		select {
		case <-stop:
			return
		case <-ticker.C:
			l.sweepGates()
		}
	}
}

File: sdks/ruby/src/lib/hatchet/worker/durable_ordered_release.rb
# frozen_string_literal: true

module Hatchet
  module WorkerRuntime
    # Ordered release of ``entry_completed`` responses for durable task
    # invocations, mixed into {DurableEventListener}.
    #
    # The engine stamps a contiguous per-task ``satisfied_order`` on durable
    # event log entries as they are satisfied. Releasing completions to user
    # code strictly in that order (gated on the previously woken continuation
    # parking again) makes the wake order deterministic, so replays reproduce
    # the originally recorded interleaving instead of racing into spurious
    # non-determinism errors.
    module DurableOrderedRelease
      # How long the ordered-release gate stays closed waiting for a woken
      # continuation to park (register its next awaited entry) before being
      # forced open with a loud warning.
      DEFAULT_PARK_TIMEOUT_SECONDS = 5.0
      # How long a hole in the satisfied-order sequence may persist (while
      # later completions are held) before the invocation's waiters are failed
      # with a NonDeterminismError instead of hanging.
      DEFAULT_GAP_TIMEOUT_SECONDS = 60.0

      # Per-invocation state serializing the release of ordered
      # ``entry_completed`` responses. Completions are released to user code in
      # ``satisfied_order``; after a release wakes a parked continuation,
      # further releases are held until that continuation parks again
      # (registers its next awaited entry), or the park timeout elapses.
      #
      # @!attribute held
      #   @return [Hash{Integer => Array(Array, Hash)}] order => [callback_key, result]
      # @!attribute released
      #   @return [Integer] highest satisfied order released so far
      # @!attribute wakes
      #   @return [Integer] continuations woken by a gated release which have
      #     not yet parked; the gate is open iff zero
      # @!attribute wake_since
      #   @return [Float, nil] monotonic time wakes last left zero
      # @!attribute gap_since
      #   @return [Float, nil] monotonic time held first became blocked on a
      #     missing order; nil when not blocked
      OrderedReleaseGate = Struct.new(:held, :released, :wakes, :wake_since, :gap_since, keyword_init: true)

      # The durable task function for the given invocation returned (or
      # otherwise has no running continuations): release any gate held on its
      # behalf so remaining ordered completions flow to late registrations.
      def notify_invocation_quiesced(durable_task_external_id, invocation_count)
        gate_key = [durable_task_external_id, invocation_count]

        @mu.synchronize do
          gate = @gates[gate_key]
          next unless gate

          gate.wakes = 0
          pump_gate(gate_key, gate)
        end
      end

      private

      # Hand a completion to a registered waiter, or buffer it for late
      # registration. Returns true if a parked continuation was woken.
      def deliver_completion(key, result)
        @mu.synchronize do
          queue = @pending_callbacks.delete(key)
          if queue
            queue << [:ok, result]
            true
          else
            @buffered_completions[key] = [Time.now, result]
            false
          end
        end
      end

      # Route an ordered completion through the invocation's gate.
      def handle_ordered_completion(gate_key, order, key, result)
        @mu.synchronize do
          gate = (@gates[gate_key] ||= OrderedReleaseGate.new(held: {}, released: 0, wakes: 0))

          if order <= gate.released
            # re-delivery of an already-released completion (e.g. after
            # reconnect): bypass the gate.
            deliver_completion(key, result)
          else
            gate.held[order] = [key, result]
            pump_gate(gate_key, gate)
          end
        end
      end

      # Release contiguously ordered completions while the gate is open.
      # Callers must hold @mu.
      def pump_gate(_gate_key, gate)
        while gate.wakes.zero?
          held = gate.held.delete(gate.released + 1)
          break unless held

          gate.released += 1

          key, result = held
          next unless deliver_completion(key, result)

          # the release woke a parked continuation: hold further releases
          # until it parks again. if nobody was waiting, the completion was
          # buffered for a continuation that is still running; keep pumping so
          # a parked continuation awaiting a later order is not deadlocked.
          gate.wakes += 1
          gate.wake_since = monotonic_now
        end

        if !gate.held.empty? && gate.wakes.zero?
          gate.gap_since ||= monotonic_now
        else
          gate.gap_since = nil
        end
      end

      # A continuation of the given invocation parked (registered its next
      # awaited entry without a buffered result): open the gate for the next
      # ordered release.
      def notify_parked(gate_key)
        @mu.synchronize do
          gate = @gates[gate_key]
          next unless gate

          gate.wakes -= 1 if gate.wakes.positive?
          pump_gate(gate_key, gate)
        end
      end

      # Enforce the park and gap timeouts on all ordered-release gates.
      def sweep_gates
        failures = []

        @mu.synchronize do
          now = monotonic_now

          @gates.keys.each do |gate_key| # rubocop:disable Style/HashEachMethods
            gate = @gates[gate_key]

            if gate.wakes.positive? && gate.wake_since && now - gate.wake_since > @park_timeout_s
              @logger&.warn(
                "durable task #{gate_key[0]} (invocation #{gate_key[1]}): continuation did not " \
                "park within #{@park_timeout_s}s after a gated release; forcing the completion " \
                "gate open. durable task code should not perform unrecorded blocking work " \
                "between durable operations",
              )
              gate.wakes = 0
              pump_gate(gate_key, gate)
            end

            next if gate.held.empty? || gate.wakes.positive? || gate.gap_since.nil?
            next unless now - gate.gap_since > @gap_timeout_s

            missing_order = gate.released + 1
            exc = Hatchet::NonDeterminismError.new(
              "completion with satisfied order #{missing_order} was never delivered while " \
              "later completions #{gate.held.keys.sort} arrived; the recorded history likely " \
              "diverged from the current code",
              task_external_id: gate_key[0],
              invocation_count: gate_key[1],
              node_id: missing_order,
            )
            @logger&.error(exc.message)

            @gates.delete(gate_key)
            failures << [gate_key, exc]
          end
        end

        failures.each { |gate_key, exc| fail_invocation_waiters(gate_key, exc) }
      end

      # Deliver an error to every pending callback and event ack belonging to
      # the given invocation.
      def fail_invocation_waiters(gate_key, exc)
        @mu.synchronize do
          @pending_callbacks.each_key do |k|
            next unless k[0] == gate_key[0] && k[1] == gate_key[1]

            @pending_callbacks.delete(k)&.<<([:err, exc])
          end

          @pending_event_acks.delete(gate_key)&.<<([:err, exc])
        end
      end

      def monotonic_now
        Process.clock_gettime(Process::CLOCK_MONOTONIC)
      end
    end
  end
end
